src/util/address.rs

```diff
@@ -388,25 +388,36 @@ impl fmt::Display for Address {
                 version: ver,
                 program: ref prog,
             } => {
-                if fmt.alternate() {
-                    //TODO format without allocation when alternate uppercase is in bech32
-                    let lower = self.to_string();
-                    write!(fmt, "{}", lower.to_ascii_uppercase())
-                } else {
-                    let hrp = match self.network {
-                        Network::Bitcoin => "bc",
-                        Network::Testnet | Network::Signet => "tb",
-                        Network::Regtest => "bcrt",
-                    };
-                    let mut bech32_writer = bech32::Bech32Writer::new(hrp, fmt)?;
-                    bech32::WriteBase32::write_u5(&mut bech32_writer, ver)?;
-                    bech32::ToBase32::write_base32(&prog, &mut bech32_writer)
-                }
+                let hrp = match self.network {
+                    Network::Bitcoin => "bc",
+                    Network::Testnet | Network::Signet => "tb",
+                    Network::Regtest => "bcrt",
+                };
+                let is_alternate = fmt.alternate();
+                let mut opt_up_writer = OptionallyUpperWriter(fmt, is_alternate);
```

Comment:
I believe dispatching here instead of using `bool` is a bit faster, but probably not something to worry about.

Reply:
Idk if that's what you mean (I think you mean putting the `if` here), but also why supply a boolean derived from `fmt` when you supply `fmt`?

Reply:
> but also why supply a boolean derived from fmt when you supply fmt?
the naive reason is that fmt is taken as something that impl fmt::Write, not as Formatter, so it's lacking `alternate()`

Reply:
@sgeisler yes, I meant putting the `if` there as was written before.

Reply:
not sure how to put the `if` as was before since something like this won't work
```
let bech32_writer = if fmt.alternate() {
  let upper =  UpperWriter(fmt);
  bech32::Bech32Writer::new(hrp, &mut upper)?;
} else {
  bech32::Bech32Writer::new(hrp, fmt)?;
}
```
and other solutions I can think about are too verbose

Reply:
Yeah, you'd have to repeat the lines below. So you're basically hand-implementing dynamic dispatch. If performance isn't a concern here (who generates thousands of QR codes per second?) then I suggest just putting `as &mut dyn fmt::Writer` there and getting rid of manual dynamic dispatch.

Reply:
Wait, wasn't your initial concern essentially avoiding a single function call in _some_ cases? _Always_ using dynamic dispatch sounds even worse (but maybe the compiler will figure it out? but the same cou). I still see that this is more elegant than having a boolen argument.

Also: we are horribly over-engineering this PR :laughing:

Reply:
I only pointed out that evaluating `if` multiple times is not great without realizing it can't be fixed without adding bloat. Thus dynamic dispatch is probably preferable. Manually-implemented dynamic dispatch has no benefit over compiler-implemented dynamic dispatch and the latter has less boilerplate, thus should be preferred.

Yeah, we're overengineering, I personally enjoy it from time to time. :smile:

Reply:
In this example https://github.com/rust-bitcoin/rust-bitcoin/pull/581#discussion_r610410437 there is also a lifetime issue not only a dispatching issue...

I am not sure how to go ahead:
1) "like it is now" -> use `OptionallyUpperWriter` with the `bool` 
2) "verbose" -> use `UpperWriter` and `if fmt.is_alternate()` with the last two lines present in both of the if branch
3) "very verbose" -> use `UpperWriter` and `OwnedWriter`
4) "inner"
```
let mut upper_writer = UpperWriter(fmt);
let writer = if upper_writer.0.alternate() {
    &mut upper_writer as &mut dyn fmt::Write
} else {
    &mut upper_writer.0 as &mut dyn fmt::Write
};
```
5) I miss what you are suggesting, please provide some code

I personally still prefer 1)

Reply:
Ah, didn't realize the lifetime issue before. There's still a useful Rust trick to fix it, made a [PR against your branch](https://github.com/RCasatta/rust-bitcoin/pull/1). :)

Reply:
very nice!